spec/Classes.tex

```diff
@@ -739,37 +1063,348 @@ \subsection{The Compiler-Generated Constructor}
+
+The second initializer implicitly invokes the parent's initializer with zero
+arguments, and then uses the parent's fields to initialize its own fields.
+\end{chapelexample}
+
+As stated earlier, the compiler will insert a zero-argument call to the
+parent's initializer if the user has not explicitly written one themselves.
+The exception to this rule is if the initializer body invokes another
+initializer on the current type. This other initializer will either contain an
+implicit or explicit call to the parent initializer, and so the calling
+initializer should not attempt to initialize the parent itself. This also means
+that parent fields may not be accessed before explicit calls to \chpl{init}.
+
+\begin{chapelexample}{superInitThisInit.chpl}
+In the following code:
+\begin{chapel}
+class Parent {
+  var x, y: real;
+}
+
+class Child : Parent {
+  var z : real;
+
+  proc init(x: real, y: real, z: real) {
+    super.init(x, y);
+    this.z = z;
+  }
+
+  proc init(z: real) {
+    this.init(0.0, 0.0, z);
+  }
+}
+
+var c = new Child(5.0);
+\end{chapel}
+\begin{chapelpost}
+writeln(c);
+\end{chapelpost}
+\begin{chapelcompopts}
+--no-warnings
+\end{chapelcompopts}
+\begin{chapeloutput}
+{x = 0.0, y = 0.0, z = 5.0}
+\end{chapeloutput}
+The second initializer does not contain an implicit call to the parent's
+initializer because it explicitly invokes another initializer.
+\end{chapelexample}
+
+\subsubsection{Calling Methods on Parent Classes}
+\label{Calling_Methods_on_Parent_Classes}
+\index{classes!initializers!initializer-inheritance!dynamic-this}
+
+Once \chpl{super.init()} returns, the dynamic type of \chpl{this} is the
+parent's type until the \chpl{complete} method is invoked (except when the
+child's fields are initialized and used). As a result, the parent's methods
+may be called and \chpl{this} may be passed to functions as the parent type.
+
+\begin{rationale}
+After \chpl{super.init()} returns the instance is in some
+partially-initialized, but valid, state. Allowing \chpl{this} to be treated as
+the parent allows for additional functionality and flexibility for users.
+\end{rationale}
+
+\begin{chapelexample}{dynamicThisInit.chpl}
+In the following code:
+\begin{chapel}
+class Parent {
+  var x, y : real;
+
+  proc foo() {
+    writeln("Parent.foo");
+  }
+}
+
+class Child : Parent {
+  var z : real;
+
+  proc init(x: real, y: real, z: real) {
+    super.init(x, y); // compiler-generated initializer
+    foo(); // Parent.foo()
+    this.z = z;
+    this.complete();
+    foo(); // Child.foo()
+  }
+
+  override proc foo() {
+    writeln("Child.foo");
+  }
+}
+
+var c = new Child(1.0, 2.0, 3.0);
+\end{chapel}
+\begin{chapelpost}
+writeln(c);
+\end{chapelpost}
+\begin{chapelcompopts}
+--no-warnings
+\end{chapelcompopts}
+Once the parent's initializer is finished, the parent method
+\chpl{foo} may be called. After the \chpl{complete} method is invoked, a call to \chpl{foo}
+resolves to the Child's overriden implementation:
```

Comment:
Is there an override section we can link to here?

Reply:
How about I don't say "overriden" at all and just say "the child's implementation" ?

Reply:
But you have to use it in your example, so you might as well link to what it means